.github/workflows/test.yml

```diff
@@ -63,17 +63,36 @@ jobs:
     - name: Run test suite
       run: tox run --skip-pkg-install -e ${{ matrix.py }}
       env:
-        COVERAGE_XML: ./coverage.xml
-    - name: Upload coverage to Codecov
-      if: matrix.os == 'ubuntu-24.04' || matrix.os == ''
-      uses: codecov/codecov-action@v5
+        COVERAGE_XML: coverage.${{ matrix.py }}.xml
+    - name: Upload .coverage file
+      uses: actions/upload-artifact@v4
       with:
-        flags: unittests
-        files: ./coverage.xml
-        fail_ci_if_error: true
-        env_vars: OS,PY,ISORT
-      env:
-        CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+        name: coverage-${{ matrix.py }}-${{ matrix.os }}-${{ matrix.isort }}-${{ matrix.black }}-${{ matrix.pydantic }}
+        path: coverage.${{ matrix.py }}.xml
+  coverage-check:
+    name: Combine Coverage Reports and Check
+    runs-on: ubuntu-latest
+    needs: test
+    steps:
+      - name: Download all coverage artifacts
+        uses: actions/download-artifact@v4
+      - name: Install the latest version of uv
+        uses: astral-sh/setup-uv@v5
+      - name: Combine coverage reports and generate XML
+        run: |
+          ls -la
+          uv run coverage combine
```

Comment:
You need to checkout the repository, afaik coverage uses the source tree to do the merge.

Reply:
You need to tell coverage what files to combine 🤔 `COVERAGE_FILE = coverage` I think will do, because then does prefix matching.

Reply:
Thank you. I'm porting the github action here.

Reply:
Yeah, was just proposing to port it to tox, and then just invoke that in GitHub action; that way is re-usabel locally 🤔 up to you though.

Reply:
I agree. Adding task makes sense since I've wanted to check integrated coverage locally.
Quick question - is it possible to create a matrix in tox not just for Python versions but also for different versions of pydantic and isort? Some of our code depends on specific library versions, so it would be helpful to run all tests in parallel locally rather than just in CI.
Plus, this would allow us to check coverage as well.

Reply:
Yeah, possible, just makes the test matrix a lot bigger. Do we want to add only what we run in the CI?

Reply:
Kinda orthogonal, but any reason we want to support all these combinations? I can understand maybe pydantic, but for black and isort I would say latest is the only supported version. How many people use these older versions?

Reply:
> I agree. Adding task makes sense since I've wanted to check integrated coverage locally.
> Quick question - is it possible to create a matrix in tox not just for Python versions but also for different versions of pydantic and isort? Some of our code depends on specific library versions, so it would be helpful to run all tests in parallel locally rather than just in CI.

So this is not possible with a lock file... because, well the two different dependency groups conflict each other. But we could not use the lock file for this case...